File: src/core/project-generator.ts
/**
 * Project generator for IDE project files.
 *
 * Generates IDE-specific project files (Visual Studio, VSCode, CLion, Xcode)
 * using UnrealBuildTool to enable development workflow integration.
 *
 * @module core/project-generator
 */

import { execa } from 'execa';
import path from 'path';
import fs from 'fs-extra';
import type { GenerateOptions, GenerateResult, IDE } from '../types/generate';
import { EngineResolver } from './engine-resolver';
import { Logger } from '../utils/logger';
import { formatError } from '../utils/error';
import { Platform } from '../utils/platform';
import { ProjectPathResolver } from './project-path-resolver';

/**
 * Represents a VSCode task definition in tasks.json.
 * Used for parsing and validating VSCode workspace task configurations.
 */
interface VSCodeTaskDefinition {
  /** The display label for the task */
  label?: string;
}

/**
 * Represents the structure of a VSCode tasks.json file.
 * Used for reading and validating VSCode workspace task configurations.
 */
interface VSCodeTasksFile {
  /** The version of the tasks file format */
  version?: string;
  /** Array of task definitions */
  tasks?: VSCodeTaskDefinition[];
}

/**
 * Type guard to check if a value is a valid VSCodeTaskDefinition.
 * Validates that the task has the expected structure with optional label property.
 * @param task - The value to check
 * @returns True if the value is a valid VSCodeTaskDefinition
 */
function isVSCodeTaskDefinition(task: unknown): task is VSCodeTaskDefinition {
  if (typeof task !== 'object' || task === null) {
    return false;
  }

  const candidate = task as VSCodeTaskDefinition;
  return candidate.label === undefined || typeof candidate.label === 'string';
}

/**
 * Type guard to check if a value is a valid VSCodeTasksFile.
 * Validates that the file has the expected structure with optional version and tasks array.
 * @param value - The value to check
 * @returns True if the value is a valid VSCodeTasksFile
 */
function isVSCodeTasksFile(value: unknown): value is VSCodeTasksFile {
  if (typeof value !== 'object' || value === null) {
    return false;
  }

  const candidate = value as VSCodeTasksFile;
  return (
    candidate.tasks === undefined ||
    (Array.isArray(candidate.tasks) && candidate.tasks.every(isVSCodeTaskDefinition))
  );
}

/**
 * Generates IDE project files for Unreal Engine projects.
 * Supports Visual Studio, VSCode, CLion, and Xcode through UBT integration.
 */
export class ProjectGenerator {
  /**
   * Generates project files for the specified IDE.
   * @param options - Generation options including IDE type and project path
   * @returns Promise resolving to generation result with list of created files
   */
  static async generate(options: GenerateOptions): Promise<GenerateResult> {
    const generatedFiles: string[] = [];
    const silent = options.silent || false;
    const logger = new Logger({ silent });

    try {
      const validatedOptions = await this.validateOptions(options);
      const { ide, projectPath, enginePath, force } = validatedOptions;

      logger.info(`Generating ${ide.toUpperCase()} project files`);
      logger.info(`Project: ${projectPath}`);
      logger.info(`Engine: ${enginePath}`);

      await this.generateWithUBT(enginePath, projectPath, force, ide, logger);

      // Collect IDE-specific generated files
      generatedFiles.push(...(await this.findGeneratedSolutionFiles(projectPath)));

      if (ide === 'vscode') {
        generatedFiles.push(...(await this.findVSCodeWorkspaceFiles(projectPath)));
        generatedFiles.push(...(await this.findVSCodeConfigFiles(projectPath)));

        const tasksFile = await this.generateVSCodeTasks(projectPath);
        if (tasksFile) {
          generatedFiles.push(tasksFile);
        }
      } else if (ide === 'clion') {
        generatedFiles.push(...(await this.findGeneratedCLionFiles(projectPath)));
      } else if (ide === 'xcode') {
        generatedFiles.push(...(await this.findGeneratedXcodeFiles(projectPath)));
      }

      return {
        success: true,
        generatedFiles,
      };
    } catch (error) {
      return {
        success: false,
        generatedFiles,
        error: formatError(error),
      };
    }
  }

  /**
   * Validates and normalizes generation options.
   * @param options - Raw generation options from user input
   * @returns Promise resolving to validated and normalized options with defaults applied
   * @throws Error if project path cannot be resolved
   * @throws Error if engine path cannot be resolved
   */
  private static async validateOptions(
    options: GenerateOptions
  ): Promise<Required<GenerateOptions>> {
    const ide: IDE = options.ide || 'sln';
    const force = options.force || false;

    const projectPath = await ProjectPathResolver.resolveOrThrow(
      options.projectPath || process.cwd()
    );

    const enginePath = await EngineResolver.resolveEnginePath({
      projectPath,
      enginePath: options.enginePath,
    });

    const silent = options.silent || false;

    return {
      ide,
      projectPath,
      enginePath,
      force,
      silent,
    };
  }

  /**
   * Generates project files using UnrealBuildTool.
   * @param enginePath - Path to the Unreal Engine installation
   * @param projectPath - Path to the .uproject file
   * @param force - Whether to force regeneration
   * @param ide - IDE type to generate files for
   * @returns Promise resolving when generation is complete
   * @throws Error if UnrealBuildTool is not found
   * @throws Error if generation fails
   */
  private static async generateWithUBT(
    enginePath: string,
    projectPath: string,
    force: boolean,
    ide: IDE,
    logger: Logger
  ): Promise<void> {
    const ubtPath = path.join(
      enginePath,
      'Engine',
      'Binaries',
      'DotNET',
      'UnrealBuildTool',
      `UnrealBuildTool${Platform.exeExtension()}`
    );

    if (!(await fs.pathExists(ubtPath))) {
      throw new Error(`UnrealBuildTool not found at: ${ubtPath}`);
    }

    const args = ['-projectfiles', `-project="${projectPath}"`, '-game', '-engine'];

    if (ide === 'vscode') {
      args.push('-VSCode');
    } else if (ide === 'clion') {
      args.push('-CLion');
    } else if (ide === 'xcode') {
      args.push('-XCodeProjectFiles');
    }

    if (force) {
      args.push('-force');
    }

    Logger.debug(`Executing: ${ubtPath} ${args.join(' ')}`);

    const childProcess = execa(ubtPath, args, {
      stdio: 'pipe',
      cwd: path.dirname(ubtPath),
    });

    if (childProcess.stdout) {
      childProcess.stdout.on('data', (data: Buffer) => {
        const output = data.toString();
        if (!output.includes('Log file created')) {
          logger.write(output);
        }
      });
    }

    if (childProcess.stderr) {
      childProcess.stderr.on('data', (data: Buffer) => {
        logger.writeError(data.toString());
      });
    }

    const result = await childProcess;

    if (!result || result.exitCode !== 0) {
      throw new Error(`Project generation failed with exit code ${result?.exitCode ?? 'unknown'}`);
    }

    Logger.success('Project files generated successfully');
  }

  /**
   * Finds generated Visual Studio solution files.
   * @param projectPath - Path to the .uproject file
   * @returns Promise resolving to array of generated solution file paths
   */
  private static async findGeneratedSolutionFiles(projectPath: string): Promise<string[]> {
    const projectDir = path.dirname(projectPath);
    const solutionFiles: string[] = [];

    const slnFiles = await fs
      .readdir(projectDir)
      .then((files) => files.filter((f) => f.endsWith('.sln')));

    solutionFiles.push(...slnFiles.map((f) => path.join(projectDir, f)));

    const filterFiles = await fs
      .readdir(projectDir)
      .then((files) => files.filter((f) => f.endsWith('.vcxproj.filters')));

    solutionFiles.push(...filterFiles.map((f) => path.join(projectDir, f)));

    const vcxprojFiles = await fs
      .readdir(projectDir)
      .then((files) => files.filter((f) => f.endsWith('.vcxproj')));

    solutionFiles.push(...vcxprojFiles.map((f) => path.join(projectDir, f)));

    return solutionFiles;
  }

  /**
   * Finds generated CLion/CMake files.
   * @param projectPath - Path to the .uproject file
   * @returns Promise resolving to array of generated CLion file paths
   */
  private static async findGeneratedCLionFiles(projectPath: string): Promise<string[]> {
    const projectDir = path.dirname(projectPath);
    const cmakePath = path.join(projectDir, 'CMakeLists.txt');
    if (await fs.pathExists(cmakePath)) {
      return [cmakePath];
    }
    return [];
  }

  /**
   * Finds generated Xcode project files.
   * @param projectPath - Path to the .uproject file
   * @returns Promise resolving to array of generated Xcode file paths
   */
  private static async findGeneratedXcodeFiles(projectPath: string): Promise<string[]> {
    const projectDir = path.dirname(projectPath);
    const entries = await fs.readdir(projectDir, { withFileTypes: true });
    const files = entries
      .filter((e) => e.isDirectory() && e.name.endsWith('.xcodeproj'))
      .map((e) => path.join(projectDir, e.name));
    return files;
  }

  /**
   * Finds generated VSCode workspace files.
   * @param projectPath - Path to the .uproject file
   * @returns Promise resolving to array of generated workspace file paths
   */
  private static async findVSCodeWorkspaceFiles(projectPath: string): Promise<string[]> {
    const projectDir = path.dirname(projectPath);
    const files = await fs
      .readdir(projectDir)
      .then((list) =>
        list.filter((f) => f.endsWith('.code-workspace')).map((f) => path.join(projectDir, f))
      );
    return files;
  }

  /**
   * Finds VSCode configuration files in the .vscode directory.
   * @param projectPath - Path to the .uproject file
   * @returns Promise resolving to array of VSCode config file paths
   */
  private static async findVSCodeConfigFiles(projectPath: string): Promise<string[]> {
    const projectDir = path.dirname(projectPath);
    const vscodeDir = path.join(projectDir, '.vscode');

    if (!(await fs.pathExists(vscodeDir))) {
      return [];
    }

    const vscodeFiles = await fs
      .readdir(vscodeDir)
      .then((files) => files.map((f) => path.join(vscodeDir, f)));

    return vscodeFiles;
  }

  /**
   * Generates or updates VSCode tasks.json with ubuild tasks.
   * @param projectPath - Path to the .uproject file
   * @returns Promise resolving to the tasks.json path, or null if not generated
   */
  private static async generateVSCodeTasks(projectPath: string): Promise<string | null> {
    const projectDir = path.dirname(projectPath);
    const vscodeDir = path.join(projectDir, '.vscode');
    const tasksPath = path.join(vscodeDir, 'tasks.json');

    await fs.ensureDir(vscodeDir);

    const tasksConfig = {
      version: '2.0.0',
      tasks: [
        {
          label: 'ubuild: Build Project',
          type: 'shell',
          command: 'ubuild',
          args: ['build'],
          group: 'build',
          problemMatcher: ['$msCompile'],
          detail: 'Build Unreal Engine project using ubuild',
        },
        {
          label: 'ubuild: Run Project',
          type: 'shell',
          command: 'ubuild',
          args: ['run', '--build-first'],
          group: 'build',
          detail: 'Build and run Unreal Engine project using ubuild',
        },
      ],
    };

    if (await fs.pathExists(tasksPath)) {
      try {
        const existingContent = await fs.readJson(tasksPath);
        if (isVSCodeTasksFile(existingContent) && existingContent.tasks) {
          existingContent.tasks = existingContent.tasks.filter(
            (task) => !(task.label && task.label.startsWith('ubuild:'))
          );
          existingContent.tasks.push(...tasksConfig.tasks);
          await fs.writeJson(tasksPath, existingContent, { spaces: 2 });
          return tasksPath;
        }
      } catch (parseError) {
        Logger.debug(
          `Failed to parse existing tasks.json, regenerating: ${formatError(parseError)}`
        );
      }
    }

    await fs.writeJson(tasksPath, tasksConfig, { spaces: 2 });
    return tasksPath;
  }
}

File: src/core/project_generator.rs
use std::path::{Path, PathBuf};
use std::process::Command;

use anyhow::{Context, Result};

use crate::types::defaults::{BUILD_CONFIG, BUILD_PLATFORM};
use crate::types::GenerateResult;
use crate::utils::file::atomic_write;
use crate::utils::logger::Logger;
use crate::utils::unreal_paths::resolve_ubt_path;

use super::compile_commands_generator::CompileCommandsGenerator;
use super::engine_resolver::EngineResolver;
use super::llvm_manager::LlvmManager;
use super::process_runner::ProcessRunner;
use super::project_path_resolver::ProjectPathResolver;

pub struct ProjectGenerator;

impl ProjectGenerator {
    pub fn generate(
        ide: &str,
        project: Option<&str>,
        engine_path: Option<&str>,
        force: bool,
    ) -> Result<GenerateResult> {
        let (project_path, engine) =
            EngineResolver::resolve_project_and_engine(project, engine_path)?;

        let project_dir = ProjectPathResolver::project_dir(&project_path);
        let ubt_path = resolve_ubt_path(&engine)?;

        let mut args = vec![
            "-projectfiles".to_string(),
            format!("-project={}", project_path.display()),
            "-game".to_string(),
            "-engine".to_string(),
        ];

        match ide {
            "vscode" => args.push("-VSCode".to_string()),
            "clion" => args.push("-CLion".to_string()),
            "xcode" => args.push("-XCodeProjectFiles".to_string()),
            _ => {} // sln / vs2022 are default
        }

        if force {
            args.push("-force".to_string());
        }

        Logger::info(&format!("Generating {ide} project files..."));

        let mut command = Command::new(&ubt_path);
        command.args(&args);
        let output = ProcessRunner::stream(&mut command)?;
        if output.exit_code != 0 {
            anyhow::bail!(
                "Project file generation failed with exit code {}",
                output.exit_code
            );
        }

        // VSCode needs a compile database and a workspace containing its editor
        // settings and tasks. Keep the compile database in the project root so
        // clangd can discover it from project source files by default.
        if ide == "vscode" {
            let llvm = LlvmManager::ensure_for_engine(&engine)?;
            CompileCommandsGenerator::generate_with_paths(
                BUILD_CONFIG,
                BUILD_PLATFORM,
                &project_path,
                &engine,
                &ubt_path,
                true,
                true,
                true,
                &[],
            )?;
            Self::generate_vscode_workspace(&project_dir, &project_path, &llvm.clangd_path)?;
        }

        let generated = Self::find_generated_files(&project_dir, ide);

        Ok(GenerateResult {
            generated_files: generated,
        })
    }

    pub fn list_ides() {
        Logger::subtitle("Available IDE types:");
        let descriptions = [
            ("sln", "Visual Studio Solution (.sln)"),
            ("vs2022", "Visual Studio 2022 project files"),
            ("vscode", "Visual Studio Code with .code-workspace config"),
            ("clion", "CLion with CMakeLists.txt"),
            ("xcode", "Xcode project files (macOS)"),
        ];
        for (name, desc) in &descriptions {
            Logger::writeln(&format!("  {name:10} {desc}"));
        }
    }

    fn find_generated_files(project_dir: &Path, ide: &str) -> Vec<PathBuf> {
        let patterns: Vec<String> = match ide {
            "sln" | "vs2022" => vec![
                format!("{}/*.sln", project_dir.display()),
                format!("{}/**/*.vcxproj", project_dir.display()),
            ],
            "clion" => vec![format!("{}/**/CMakeLists.txt", project_dir.display())],
            "xcode" => vec![format!("{}/**/*.xcodeproj", project_dir.display())],
            "vscode" => vec![
                format!("{}/*.code-workspace", project_dir.display()),
                format!("{}/compile_commands.json", project_dir.display()),
                format!("{}/.vscode/**", project_dir.display()),
            ],
            _ => Vec::new(),
        };

        let mut files = Vec::new();
        for pattern in &patterns {
            if let Ok(entries) = glob::glob(pattern) {
                files.extend(entries.filter_map(Result::ok));
            }
        }
        files
    }

    fn generate_vscode_workspace(
        project_dir: &Path,
        project_path: &Path,
        clangd_path: &Path,
    ) -> Result<()> {
        let workspace_path = Self::vscode_workspace_path(project_dir, project_path);
        let mut workspace = if workspace_path.exists() {
            Self::read_json_object(&workspace_path)?
        } else {
            serde_json::Map::new()
        };

        workspace
            .entry("folders".to_string())
            .or_insert_with(|| serde_json::json!([{ "path": "." }]));

        let mut settings = Self::workspace_object(&workspace, "settings")?;
        settings.extend([
            (
                "C_Cpp.default.compileCommands".to_string(),
                serde_json::json!("${workspaceFolder}/compile_commands.json"),
            ),
            (
                "C_Cpp.intelliSenseEngine".to_string(),
                serde_json::json!("disabled"),
            ),
            (
                "clangd.arguments".to_string(),
                serde_json::json!([
                    "--compile-commands-dir=${workspaceFolder}",
                    "--background-index",
                    "--j=8",
                    "--index-store-path=.clangd/index",
                    "--pch-storage=disk",
                    "--limit-results=200",
                    "--header-insertion=iwyu"
                ]),
            ),
            (
                "clangd.path".to_string(),
                serde_json::json!(clangd_path.to_string_lossy()),
            ),
        ]);
        workspace.insert("settings".to_string(), serde_json::Value::Object(settings));

        let generated_tasks = serde_json::json!({
            "version": "2.0.0",
            "tasks": [
                {
                    "label": "ubuild: Build Project",
                    "type": "shell",
                    "command": "ubuild",
                    "args": ["build"],
                    "group": "build",
                    "problemMatcher": ["$msCompile"],
                    "detail": "Build Unreal Engine project using ubuild"
                },
                {
                    "label": "ubuild: Run Project",
                    "type": "shell",
                    "command": "ubuild",
                    "args": ["run", "--build-first"],
                    "group": "build",
                    "detail": "Build and run Unreal Engine project using ubuild"
                }
            ]
        });
        let mut tasks = Self::workspace_object(&workspace, "tasks")?;
        Self::merge_generated_tasks(&mut tasks, &generated_tasks)?;

        workspace.insert("tasks".to_string(), serde_json::Value::Object(tasks));

        let json = serde_json::to_string_pretty(&workspace)?;
        atomic_write(&workspace_path, json)?;

        Ok(())
    }

    fn read_json_object(path: &Path) -> Result<serde_json::Map<String, serde_json::Value>> {
        let content = std::fs::read_to_string(path)
            .with_context(|| format!("Failed to read {}", path.display()))?;
        serde_json::from_str(&content)
            .with_context(|| format!("Failed to parse JSON object {}", path.display()))
    }

    fn workspace_object(
        workspace: &serde_json::Map<String, serde_json::Value>,
        key: &str,
    ) -> Result<serde_json::Map<String, serde_json::Value>> {
        match workspace.get(key) {
            None => Ok(serde_json::Map::new()),
            Some(serde_json::Value::Object(value)) => Ok(value.clone()),
            Some(_) => anyhow::bail!("VSCode workspace field '{key}' must be an object"),
        }
    }

    fn merge_generated_tasks(
        tasks: &mut serde_json::Map<String, serde_json::Value>,
        generated: &serde_json::Value,
    ) -> Result<()> {
        let generated_tasks = generated
            .get("tasks")
            .and_then(serde_json::Value::as_array)
            .ok_or_else(|| anyhow::anyhow!("Generated VSCode tasks are not an array"))?;
        let task_list = tasks
            .entry("tasks".to_string())
            .or_insert_with(|| serde_json::Value::Array(Vec::new()));
        let Some(task_list) = task_list.as_array_mut() else {
            anyhow::bail!("VSCode workspace field 'tasks.tasks' must be an array");
        };

        for generated_task in generated_tasks {
            let generated_label = generated_task
                .get("label")
                .and_then(serde_json::Value::as_str);
            if let Some(label) = generated_label {
                if let Some(existing) = task_list.iter_mut().find(|task| {
                    task.get("label")
                        .and_then(serde_json::Value::as_str)
                        .is_some_and(|existing_label| existing_label == label)
                }) {
                    *existing = generated_task.clone();
                    continue;
                }
            }
            task_list.push(generated_task.clone());
        }
        tasks.insert("version".to_string(), serde_json::json!("2.0.0"));
        Ok(())
    }

    fn vscode_workspace_path(project_dir: &Path, project_path: &Path) -> PathBuf {
        project_dir.join(format!(
            "{}.code-workspace",
            ProjectPathResolver::project_name(project_path)
        ))
    }
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::Path;

    use anyhow::Result;
    use tempfile::tempdir;

    use super::ProjectGenerator;

    #[test]
    fn embeds_vscode_settings_and_tasks_in_workspace() -> Result<()> {
        let directory = tempdir()?;
        let project_path = directory.path().join("Game.uproject");
        let workspace_path = directory.path().join("Game.code-workspace");
        fs::write(&project_path, "{}")?;
        fs::write(
            &workspace_path,
            r#"{
                "folders": [{"path": "."}],
                "settings": {"custom.setting": true},
                "extensions": {"recommendations": ["llvm-vs-code-extensions.vscode-clangd"]}
            }"#,
        )?;

        ProjectGenerator::generate_vscode_workspace(
            directory.path(),
            &project_path,
            Path::new("clangd"),
        )?;

        let content = fs::read_to_string(&workspace_path)?;
        let workspace: serde_json::Value = serde_json::from_str(&content)?;
        assert_eq!(workspace["settings"]["custom.setting"], true);
        assert_eq!(
            workspace["settings"]["C_Cpp.intelliSenseEngine"],
            "disabled"
        );
        assert_eq!(workspace["tasks"]["version"], "2.0.0");
        assert!(workspace["tasks"]["tasks"].is_array());
        assert_eq!(
            workspace["extensions"]["recommendations"][0],
            "llvm-vs-code-extensions.vscode-clangd"
        );
        assert!(!directory.path().join(".vscode/settings.json").exists());
        assert!(!directory.path().join(".vscode/tasks.json").exists());
        Ok(())
    }

    #[test]
    fn creates_project_workspace_when_ubt_did_not_create_one() -> Result<()> {
        let directory = tempdir()?;
        let project_path = directory.path().join("Game.uproject");
        fs::write(&project_path, "{}")?;

        ProjectGenerator::generate_vscode_workspace(
            directory.path(),
            &project_path,
            Path::new("clangd"),
        )?;

        let workspace_path = directory.path().join("Game.code-workspace");
        let content = fs::read_to_string(workspace_path)?;
        let workspace: serde_json::Value = serde_json::from_str(&content)?;
        assert_eq!(workspace["folders"][0]["path"], ".");
        Ok(())
    }
}
